hera_cal/nucal.py

```diff
@@ -0,0 +1,473 @@
+            ants=ants,
+            ex_ants=ex_ants,
+            pols=pols,
+            ex_pols=ex_pols,
+        )
+        if len(_baselines) == 0:
+            self._baselines = []
+            self.baseline_lengths = []
+        else:
+            new_bls = []
+            new_bls_lengths = []
+            for bls in _baselines[0]:
+                index = self._baselines.index(bls)
+                if min_bl_cut is not None and self.baseline_lengths[index] < min_bl_cut:
+                    continue
+                if max_bl_cut is not None and self.baseline_lengths[index] > max_bl_cut:
+                    continue
+                new_bls.append(bls)
+                new_bls_lengths.append(self.baseline_lengths[index])
+
+            self._baselines = new_bls
+            self.baseline_lengths = new_bls_lengths
+
+    def __iter__(self):
+        """Iterate through baselines in the radially redundant group"""
+        return iter(self._baselines)
+
+    def __len__(self):
+        """Return the length of the baselines list"""
+        return len(self._baselines)
+
+    def __getitem__(self, index):
+        """Get the baseline at the chosen index"""
+        return self._baselines[index]
+
+
+class FrequencyRedundancy:
+    """List-like object that contains groups RadialRedundantGroup objects.
+    Functions similarly to the output of redcal.get_reds for frequency redundant
+    calibration. In addition to mimicking list functionality, this object also filters
+    radially redundant groups based on a number of factors, can get specific polarizations, and
+    radially redundant and spatially redundant groups by baseline key.
+    """
+
+    def __init__(
+        self, antpos, reds=None, blvec_error_tol=1e-9, pols=["nn"], bl_error_tol=1.0
+    ):
+        """
+        Parameters:
+        ----------
+        antpos : dict
+            Antenna positions in the form {ant_index: np.array([x,y,z])}.
+        reds : list of list
+            List of lists of baseline keys. Can be determined using redcal.get_reds
+        pols : list of strs
+            List of polarization strings to be used in the frequency redundant group
+        """
+        self.antpos = antpos
+
+        if reds is None:
+            reds = redcal.get_reds(antpos, pols=pols, bl_error_tol=bl_error_tol)
+
+        self._mapped_reds = {red[0]: red for red in reds}
+
+        self._radial_groups = get_unique_orientations(
+            antpos, reds=reds, pols=pols, blvec_error_tol=blvec_error_tol
+        )
+
+    def get_radial_group(self, key):
+        """
+        Get baselines with the same radal heading as a given baseline
+
+        Parameters:
+        ----------
+        key : tuple
+            Baseline key of type (ant1, ant2, pol)
+
+        Returns:
+        -------
+        group : list of tuples
+            List of baseline tuples that have the same radial headings
+
+        """
+        # Identify headings
+        for group_key in self._mapped_reds:
```

Comment:
Not great that you have to do this loop every time you give it a single key. Seems to me like there should be a dictionary under the hood that figures this out once and for all.